core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,1155 @@
+/// for safe addition.
+impl ops::Add for Fixed64 {
+	type Output = Self;
+
+	fn add(self, rhs: Self) -> Self::Output {
+		Self(self.0 + rhs.0)
+	}
+}
+
+/// Note that this is a standard, _potentially-panicking_, implementation. Use `Saturating` trait
+/// for safe subtraction.
+impl ops::Sub for Fixed64 {
+	type Output = Self;
+
+	fn sub(self, rhs: Self) -> Self::Output {
+		Self(self.0 - rhs.0)
+	}
+}
+
+impl CheckedSub for Fixed64 {
+	fn checked_sub(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_sub(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+impl CheckedAdd for Fixed64 {
+	fn checked_add(&self, rhs: &Self) -> Option<Self> {
+		if let Some(v) = self.0.checked_add(rhs.0) {
+			Some(Self(v))
+		} else {
+			None
+		}
+	}
+}
+
+/// Helper gcd function used in Rational128 implementation.
+fn gcd(a: u128, b: u128) -> u128 {
+	match ((a, b), (a & 1, b & 1)) {
+		((x, y), _) if x == y => y,
+		((0, x), _) | ((x, 0), _) => x,
+		((x, y), (0, 1)) | ((y, x), (1, 0)) => gcd(x >> 1, y),
+		((x, y), (0, 0)) => gcd(x >> 1, y >> 1) << 1,
+		((x, y), (1, 1)) => {
+			let (x, y) = (min(x, y), max(x, y));
+			gcd((y - x) >> 1, x)
+		},
+		_ => unreachable!(),
+    }
+}
+
+/// A wrapper for any rational number with a 128 bit numerator and denominator.
+#[derive(Clone, Copy, Default)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Rational128(u128, u128);
+
+/// Some helper functions to work with 128bit numbers. Note that the functionality provided here is
+/// only sensible to use with 128bit numbers because for smaller sizes, you can always rely on
+/// assumptions of a bigger type (u128) being available, or simply create a per-thing and use the
+/// multiplication implementation provided there.
+pub mod helpers_128bit {
+	use super::Perquintill;
+	use crate::traits::Zero;
+
+	/// return the number of bits that are needed to store `n`.
+	pub(crate) fn bits_of(n: u128) -> u32 {
+		128 - n.leading_zeros()
+	}
+
+	/// Tries to compute `a * b / c`. The approach is:
+	///   - Simply try `a * b / c`.
+	///   - Else: Swap the operations (divide first) if `a > c` (division is possible) and `b <= c`
+	///     (overflow cannot happen)
+	///   - Else: Try and compute a perquintillion from `b / c` if `b <= c`.
+	///   - Else: return `Err`.
+	///
+	/// If none worked then return Error.
+	///
+	/// c must be greater than or equal to 1.
+	pub fn multiply_by_rational(a: u128, b: u128, c: u128) -> Result<u128, &'static str> {
+		if a.is_zero() { return Ok(Zero::zero()); }
+		let c = c.max(1);
+
+		if let Some(x) = a.checked_mul(b) {
+			// This is the safest way to go. Try it.
+			Ok(x / c)
+		} else if a > c && b <= c {
+			// if it can be safely swapped and it is a fraction, then swap. The operations in this
+			// block cannot overflow since b <= c.
+			let q = a / c;
+			let r = a % c;
+			let r_additional = (r.saturating_mul(b)) / c;
```

Comment:
I think this saturation can be an issue for instance for this call: `let res = multiply_by_rational(2u128.pow(66) - 1, 2u128.pow(65) - 1, 2u128.pow(65)).unwrap();`

see the error is about 50% of the result
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=87c04c91fbeed659c17d0f40abfe242d

Reply:
yeah -- the statement `r * b / c;` is basically a recursive situation of the same function: the result is for sure less than `r` but we cannot compute in size... darn. 

Though, using this works perfectly. Potentially lossy and not fast but accurate. 

```suggestion
			let r_additional = multiply_by_rational_greedy(r, b, c);
```

note: the must_be in the gist you send is `73786976294838206461`.

Reply:
oh yes.
Yes ok multiply_by_rational can be greedy now maybe we should only implement greedy version, is there a use for not greedy one ?

Reply:
Only in `lcm()` and `to_den()` where we don't have the guarantee that `b < c` and we might overflow actually. in other phragmen use cases, we know that `b < c`. This is because the greedy one just returns `a` if the result is bigger than u128 and we never want this when computing lcm and such,

Reply:
Other than that the non-greedy one is just a nice way to put some code away. The greedy always tries the normal one and if it fails starts the loop of accuracy reduction.

Reply:
well I think I'm still against having the non greedy calling greedy in one of its branch.

also maybe we could just remove this branch and use perquintill approximation.

Reply:
> also maybe we could just remove this branch and use perquintill approximation.

As you pointed out in some examples, this is poorly inaccurate in many cases. The general approach should be to use the divide and multiply (swap) and if that fails, try perquintillion. Basically the second branch (swap) is much more accurate and preferable. For example

```rust
assert_eq!(
	Perquintill::from_rational_approximation(2u128, 3) * MAX128,
	multiply_by_rational(MAX128, 2u128, 3).unwrap(),
	"should have been = {}", mul_div(MAX128, 2, 3),
);
```

will fail with 

```bash
thread 'sr_arithmetic::test_rational128::multiply_by_rational_greedy_works' panicked at 'assertion failed: `(left == right)`
  left: `226854911280625642082061493673886498661`,
 right: `226854911280625642308916404954512140970`: should have been = 226854911280625642308916404954512140970'
```